// @contentlayer/source-files/dist/index.js
import * as tracing_1 from "@effect-ts/core/Tracing";
const fileName_1 = "packages/@contentlayer/source-files/src/index.ts";
import { processArgs, SourceProvideSchemaError } from '@contentlayer/core';
import { unknownToPosixFilePath } from '@contentlayer/utils';
import { pipe, T } from '@contentlayer/utils/effect';
import { fetchData } from './fetchData/index.js';
import { makeCoreSchema } from './schema/provideSchema.js';
export * from './types.js';
export * from './schema/defs/index.js';
export const makeSource = async (args) => {
    const { options, extensions, restArgs: { documentTypes, contentDirPath, contentDirInclude, contentDirExclude, onUnknownDocuments = 'skip-warn', onMissingOrIncompatibleData = 'skip-warn', onExtraFieldData = 'warn', }, } = await processArgs(args);
    const flags = { onUnknownDocuments, onExtraFieldData, onMissingOrIncompatibleData };
    const documentTypeDefs = (Array.isArray(documentTypes) ? documentTypes : Object.values(documentTypes)).map((_) => _.def());
    return {
        type: 'local',
        extensions: extensions ?? {},
        options,
        provideSchema: (esbuildHash) => (T.mapError_(makeCoreSchema({ documentTypeDefs, options, esbuildHash }), (error) => new SourceProvideSchemaError({ error }), fileName_1 + ":104:19")),
        fetchData: ({ schemaDef, verbose }) => fetchData({
            coreSchemaDef: schemaDef,
            documentTypeDefs,
            flags,
            options,
            contentDirPath: unknownToPosixFilePath(contentDirPath),
            contentDirExclude: (contentDirExclude ?? contentDirExcludeDefault).map((_) => unknownToPosixFilePath(_)),
            contentDirInclude: (contentDirInclude ?? []).map((_) => unknownToPosixFilePath(_)),
            verbose,
        }),
    };
};
export const contentDirExcludeDefault = ['node_modules', '.git', '.yarn', '.cache', '.next', '.contentlayer'];
